### PR TITLE
DataTable: fix sortable column accessibility issue (aria-sort, discoverability, sort status)

### LINKED ISSUE #7924: DataTable: sortable columns missing aria-sort="none" and unsorted icon should be aria-hidden
### Is this something you or your team is interested in contributing a pull request towards?

Yes

### Expected Behavior
Two related fixes needed for sortable columns that are not currently sorted:

The <th> element should have aria-sort="none" so AT users can discover the column is sortable before interacting with it. Currently only "ascending" and "descending" are set in Header.js the unsorted state is omitted entirely.

The neutral ↕ sort indicator icon shown on unsorted sortable columns should have aria-hidden="true". It is decorative in this context sortability will be communicated via aria-sort="none". Without aria-hidden, some AT may announce an unlabelled SVG since iconAriaLabel is undefined in this state.

### Actual Behavior

aria-sort is absent on sortable columns that are not currently sorted
The ↕ icon on unsorted sortable columns has no aria-hidden="true" and no label

### URL, Screenshot, or CodeSandbox

n/a

### Steps to Reproduce

Render a DataTable with sortable columns. Inspect the <th> elements on unsorted columns — no aria-sort attribute is present and the sort icon SVG has no aria-hidden.

https://storybook.grommet.io/?path=/story/visualizations-datatable-sort--sort

### Your Environment

latest Grommet

### PR DESCRIPTION
**Pull request by Sulaymon333**
#### What does this PR do?

Fixes [#7924](https://github.com/grommet/grommet/issues/7924) by combining and reconciling the two independent fixes proposed in [#7934](https://github.com/grommet/grommet/pull/7934) and [#7969](https://github.com/grommet/grommet/pull/7969), which each solved part of the problem but had unresolved trade-offs when reviewed together.

**Specifically:**
- Sortable-but-unsorted columns now get `aria-sort="none"` (was previously omitted entirely).
- The sort indicator icon is now always `aria-hidden` (purely decorative) — no more unlabeled/inconsistently-labeled SVG exposed to assistive tech.
- Sortable column header buttons now expose a full, descriptive accessible name in **every** state (not just when active), e.g.:
  - Unsorted: *"A sortable, activate to sort ascending"*
  - Sorted ascending: *"A sorted ascending, activate to sort descending"*
  - Sorted descending: *"A sorted descending, activate to sort ascending"*
- This is implemented via a visually-hidden text node inside the button (`HiddenText`, same pattern already used in Cards.js) rather than a static `aria-label` override on the button — deliberately keeping the accessible name **content-derived**, which preserves the pre-existing, verified-working real-time announcement behavior on sort activation. 
- 3 new message keys (`dataTable.sortable`, `sortedAscending`, `sortedDescending`) 
- `dataTable.ascending`/`dataTable.descending` (the old icon-label keys) are **not removed** — they're no longer read anywhere, but kept in `propTypes.js/index.d.ts/default.json` and marked `@deprecated` in the type declarations, to avoid a breaking change for consumers who may have translated/overridden them. This needs further team's agreement/decision to remove or not if decided.

#### Where should the reviewer start?

`Header.js` — specifically the sortable-column block where `ariaSort`, the `HiddenText` span, and the message-id selection logic live. Then `default.json` for the new copy.

## What testing has been done on this PR?

- Full DataTable-test.tsx suite, and updated snapshot.
- Added a `jest-axe` accessibility check (`toHaveNoViolations`) on a sorted DataTable — zero violations.
- Added/updated tests asserting `aria-sort` transitions and the exact accessible name in all three states (unsorted/ascending/descending), plus a non-sortable-column and a themeless-icon edge case.
- **Manually tested live with VoiceOver** on the actual Sort story (`visualizations-datatable-sort--sort`, the same story referenced in the original ticket): confirmed Tab-in discoverability on both unsorted and already-sorted columns, and confirmed the real-time announcement on activation (pressing Enter) still fires correctly.

#### How should this be manually tested?

1. Open the **Visualizations → DataTable → Sort** story.
2. With VoiceOver (or NVDA/JAWS) on, Tab to an unsorted column header — confirm it announces column name + "sortable, activate to sort ascending".
3. Press Enter — confirm the sort direction change is announced immediately.
4. Tab away and back to that now-sorted column — confirm it announces the current direction + the alternate action.

#### Do Jest tests follow these best practices?
- [x] `screen` is used for querying.
- [x] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [x] `asFragment()` is used for snapshot testing.

## Any background context you want to provide?

This PR was is a combined solution that merges #7934 and #7969, since each had a piece the other was missing: #7934's `aria-sort="none"` fix was working but its discoverability copy was minimal; #7969's richer copy was great but its implementation mechanism (static `aria-label` override) reproduced a regression where the sorting action was not announced via regular `Enter` key as we have in production today. `AnnounceContext`/`aria-live` was not used in this PR implementation since it was not the direction the team would like to go.

#### What are the relevant issues?
Closes #7924. Once this PR is approved and merged, both #7934 and #7969 both should be closed in favor of this PR.

#### Screenshots (if appropriate)
N/A (screen-reader-only change; no visual difference).

## Do the grommet docs need to be updated?
Yes — the `messages` prop documentation for `DataTable` should list the 3 new keys and note the 2 deprecated ones.

## Should this PR be mentioned in the release notes?
Yes — it's an accessibility fix for a real reported issue.

#### Is this change backwards compatible or is it a breaking change?
 Mostly backwards compatible, with one specific exception. No props removed or renamed; `dataTable.ascending`/`descending` are retained (deprecated, not deleted) specifically to avoid breaking existing `messages` overrides or TypeScript consumers but no longer used. This needs further team's agreement/decision to remove or not if decided.

### CODE REVIEW

File: src/js/components/DataTable/Header.js [hunk 1]
```diff
@@ -359,29 +404,33 @@ const Header = forwardRef(
...
+                }
+
+                // sort status + alternate action text, exposed via a
+                // visually hidden span rather than an aria-label override on
+                // the button. This keeps the button's accessible name
+                // content-derived (visible header text + this hidden text),
+                // which is the same mechanism that already reliably
+                // announces sort changes in real time on activation -
+                // unlike a static aria-label override, which does not.
+                let hiddenTextId = 'dataTable.sortable';
```
Comment (Sulaymon333) [inline src/js/components/DataTable/Header.js:432]: `dataTable.ascending`/`dataTable.descending` (previously used as the sort icon's aria-label) are no longer read anywhere in this component - superseded by `sortable`/`sortedAscending`/`sortedDescending` below. Left declared in propTypes/index.d.ts/default.json to avoid a breaking change for consumers who may have overridden them for translation; open question for the team on whether/when to formally deprecate and remove.

File: src/js/components/DataTable/Header.js [hunk 2]
```diff
@@ -315,6 +330,36 @@ const Header = forwardRef(
               size,
               units,
             }) => {
+              // Extract a plain-text label for the column, used for
+              // accessible names (e.g. resizer, sortable button a11yTitle).
+              // Falls back to the column property when no text can be
+              // derived from a non-string / node header.
```
[the remark this answers is not in the document]
Comment (Sulaymon333) [reply src/js/components/DataTable/Header.js]: addressed via this [commit](https://github.com/grommet/grommet/pull/7988/changes/12c2ddfe69b5d2719935605ff9713c204f79c465)

File: src/js/components/DataTable/__tests__/DataTable-test.tsx [hunk 3]
```diff
@@ -178,6 +201,139 @@ describe('DataTable', () => {
...
+        name: 'B sortable, activate to sort ascending',
+      }),
+    ).toBeInTheDocument();
+
+    // once a column becomes the active sort, aria-sort reflects direction
+    // and the button's accessible name (derived from content: visible
+    // header text + a visually hidden status/action span) updates to
+    // match, preserving the existing real-time sort announcement behavior
+    // since nothing is overridden via a static aria-label.
+    fireEvent.click(getByText('A'));
```
[the remark this answers is not in the document]
Comment (Sulaymon333) [reply src/js/components/DataTable/__tests__/DataTable-test.tsx]: addressed via this [commit](https://github.com/grommet/grommet/pull/7988/changes/12c2ddfe69b5d2719935605ff9713c204f79c465)

File: src/js/components/DataTable/Header.js (hunk 1 above)
Comment (britt6612) [reply src/js/components/DataTable/Header.js:432]: so if someone has something declared for their `dataTable.ascending/dataTable.descending` that wont be set anywhere anymore? I feel like that would be a problem

File: src/js/components/DataTable/index.d.ts [hunk 4]
```diff
@@ -87,10 +87,24 @@ export interface DataTableProps<TRowType = any> {
   gridArea?: GridAreaType;
   margin?: MarginType;
   messages?: {
+    /**
+     * @deprecated no longer read by DataTable's sort UI as of the #7924
+     * accessibility fix - sort status is now communicated via `sortable`/
+     * `sortedAscending`/`sortedDescending` below. Kept to avoid a breaking
+     * change for existing consumers; pending team decision on formal
```
Comment (britt6612) [inline src/js/components/DataTable/index.d.ts]: I think instead of the comment in line for the code we need to discuss as a team. Can you plan on setting up time or send out a message for everyone?

**Comment by britt6612** [comment]
@sulaymon-tajudeen-hpe 

The PR description calls this "backwards compatible" because no props are removed, but functionally it breaks any consumer who were using something like this

```
 messages={{
    ascending: 'Trié par ordre croissant',
    descending: 'Trié par ordre décroissant',
  }}

```

**Comment by Sulaymon333** [comment]
> @sulaymon-tajudeen-hpe
> 
> The PR description calls this "backwards compatible" because no props are removed, but functionally it breaks any consumer who were using something like this
> 
> ```
>  messages={{
>     ascending: 'Trié par ordre croissant',
>     descending: 'Trié par ordre décroissant',
>   }}
> ```

Good catch, and you're right. I have expanded on the "backwards compatible" in the description to be more precise to clarify about what is actually preserved vs. not:

**- Prop/type surface:** unchanged. No props removed, nothing breaks at compile time, `messages.dataTable.ascending/descending` are still valid, typed properties (marked @deprecated with an explanation in the `.d.ts` files).
**- Functional behavior:** broken for that specific customization. Neither key is read anywhere in the sort UI anymore. similar implementation on #7969 — the accessible sort status/action text now comes entirely from three new keys (`dataTable.sortable`, `dataTable.sortedAscending, dataTable.sortedDescending). Anyone who set ascending/descending specifically to translate the old icon label will find that translation silently stops applying, with no error or warning. Updating chanegelog and providing migration guide will help solve this.

I have considered splicing/joining old translated values into the new sentence structure as a fallback, but ruled it out since the old keys were short fragments ("ascending"/"descending" used as an SVG label and supplying full sentence on SVG label feels quite weird), while the new ones are full sentences with an embedded action hint ("activate to sort ascending"). There's no safe way to graft a translator's short fragment into the new sentence shape across arbitrary languages without risking broken grammar in some locales — that felt like opening another can of potential bug. Open to other possible approach to honor `dataTable.ascending/dataTable.descending` without possible issues.

So: keeping the deprecated-but-present keys (no hard break for existing code), We should be updating a changelog/migration note to be explicit that this is a breaking change specifically for consumers customizing `messages.dataTable.ascending/descending` and those will not be applied.

Migration note sample change (see note below).


```js
// Before
messages={{
  dataTable: {
    ascending: 'Trié par ordre croissant',
    descending: 'Trié par ordre décroissant',
  },
}}

// After
messages={{
  dataTable: {
    sortable: 'triable, activer pour trier par ordre croissant',
    sortedAscending: 'trié par ordre croissant, activer pour trier par ordre décroissant',
    sortedDescending: 'trié par ordre décroissant, activer pour trier par ordre croissant',
  },
}}
```

File: src/js/components/DataTable/index.d.ts (hunk 4 above)
Comment (Sulaymon333) [reply src/js/components/DataTable/index.d.ts]: Sure, this will be part of next dev chat. The comment will be updated after the dev chat/team decision.

File: src/js/components/DataTable/Header.js (hunk 1 above)
Comment (Sulaymon333) [reply src/js/components/DataTable/Header.js:432]: addressed in the [comment below](https://github.com/grommet/grommet/pull/7988#issuecomment-4992062389)

**Comment by britt6612** [comment]
You're right that this mirrors #7969 I went and checked, and that PR added `sortable`, `sortedAscending`, `sortedDescending`, and `sortButton` to `default.json` and switched to a composed `sortButton` message, while leaving the old `ascending`/`descending` keys just as unread. I hadn't connected that until now.

On the "no safe way to honor `ascending`/`descending`" point  there might actually be a middle ground worth considering: instead of adding two new keys (`sortedAscending`/`sortedDescending`) and leaving `ascending`/`descending` permanently deprecated and unread, we could repurpose `ascending`/`descending` themselves to hold the full sentence (what `sortedAscending`/`sortedDescending` currently hold), and only add one new key, `sortable`, for the unsorted state (which never had a key before).

**Pros:** existing customizations of `ascending`/`descending` get rendered again instead of silently dropped it directly answers the original concern rather than just documenting around it and it only adds one new key instead of two, with no permanently-dead deprecated props to carry forward.

**Cons:** it doesn't fully solve the underlying problem, just relocates it an existing translation like `'Trié par ordre croissant'` would render as the full hidden text but without the new action hint, since we're not grafting fragments together. So the actual accessibility improvement still doesn't reach un-migrated translations until the consumer updates their string. There's also a minor semantic-drift risk: `ascending` used to mean "the short word," and now would mean "a full sentence with an action hint"  a future translator trusting the key name without reading the migration note could plausibly regress back to a short fragment.

Neither option is perfect one preserves old behavior but leaves an a11y gap for un-migrated consumers, the other cleanly delivers the a11y fix but silently drops customizations (same as #7969 did). Given that tradeoff, and that this exact pattern has come up twice, I think it's worth pulling in a few other maintainers for input before we settle on final key names/behavior curious if others have opinions.

@grommet/devs

**Comment by Sulaymon333** [comment]
> You're right that this mirrors #7969 I went and checked, and that PR added `sortable`, `sortedAscending`, `sortedDescending`, and `sortButton` to `default.json` and switched to a composed `sortButton` message, while leaving the old `ascending`/`descending` keys just as unread. I hadn't connected that until now.
> 
> On the "no safe way to honor `ascending`/`descending`" point there might actually be a middle ground worth considering: instead of adding two new keys (`sortedAscending`/`sortedDescending`) and leaving `ascending`/`descending` permanently deprecated and unread, we could repurpose `ascending`/`descending` themselves to hold the full sentence (what `sortedAscending`/`sortedDescending` currently hold), and only add one new key, `sortable`, for the unsorted state (which never had a key before).
> 
> **Pros:** existing customizations of `ascending`/`descending` get rendered again instead of silently dropped it directly answers the original concern rather than just documenting around it and it only adds one new key instead of two, with no permanently-dead deprecated props to carry forward.
> 
> **Cons:** it doesn't fully solve the underlying problem, just relocates it an existing translation like `'Trié par ordre croissant'` would render as the full hidden text but without the new action hint, since we're not grafting fragments together. So the actual accessibility improvement still doesn't reach un-migrated translations until the consumer updates their string. There's also a minor semantic-drift risk: `ascending` used to mean "the short word," and now would mean "a full sentence with an action hint" a future translator trusting the key name without reading the migration note could plausibly regress back to a short fragment.
> 
> Neither option is perfect one preserves old behavior but leaves an a11y gap for un-migrated consumers, the other cleanly delivers the a11y fix but silently drops customizations (same as #7969 did). Given that tradeoff, and that this exact pattern has come up twice, I think it's worth pulling in a few other maintainers for input before we settle on final key names/behavior curious if others have opinions.
> 
> @grommet/devs

@britt6612 This sounds like a good middle ground - repurposing `dataTable.ascending`/`dataTable.descending` instead of adding new keys avoids the breaking-change problem.

**Quick bit of context:** in production today, the sort icon carries a short `aria-label` (e.g. "ascending"). In this PR, the icon is unconditionally `aria-hidden` and purely decorative - the accessible text moved into a dedicated visually-hidden `HiddenText` span. Repurposing `ascending`/`descending` only changes which message feeds that span; it doesn't touch the icon at all.

**On the trade-off**:

- **Repurpose `ascending`/`descending`** (this proposal): consumers who already customized these get their translation back immediately on upgrade, no code change needed — matching `master`'s baseline behavior. **Cost:** they won't get the new "activate to sort X" hint until they update their string. A regression to *pre-PR* behavior, not a new break.
- **This PR as it stands**: the a11y improvement reaches everyone immediately, but anyone who customized `ascending`/`descending` on `master` finds it silently unread after upgrading — a real regression from `master`, and the same failure mode [#7969](https://github.com/grommet/grommet/pull/7969) has.

**On the semantic-drift risk** (key `ascending`/`descending` quietly changing meaning from "a word" to "a full sentence with an action hint"). This is addressable rather than a blocker: update the type/JSDoc comment on `ascending`/`descending` to describe the new expected shape explicitly, and update the default value in default.json to model a full sentence rather than a bare word - so anyone using the default as a translation starting point copies the right pattern instead of inferring a short word from the key name.

Agreed on looping in other maintainers for wider perspective before locking this down.

**Comment by Sulaymon333** [comment]
Latest commits, preserve the original `ascending/descending` keys as discussed above based on team decision. @britt6612 @jcfilben, Can you have another round of review? Thanks

**Comment by britt6612** [comment]
> Latest commits, preserve the original `ascending/descending` keys as discussed above based on team decision. @britt6612 @jcfilben, Can you have another round of review? Thanks

I think thats the best option given its enhancing the accessibility thank you for preserving this original `ascending/descending` keys.

File: src/js/languages/default.json [hunk 5]
```diff
@@ -62,11 +62,11 @@
     "totalSingle": "{total} {items}"
   },
   "dataTable": {
-    "ascending": "ascending",
+    "ascending": "sorted ascending, activate to sort descending",
```
Comment (britt6612) [inline src/js/languages/default.json:80]: **Thought on the VO experience:** right now the full `"sorted [direction], activate to sort [other]"` sentence announces every time you focus a sorted column, not just the first time.

Since `Header.js` already branches the hidden text by state what if we kept the full hint only for the **unsorted** state (useful for discoverability) and trimmed the sorted strings to just the state?

| key | current | proposed |
|---|---|---|
| `ascending` | `"sorted ascending, activate to sort descending"` | `"sorted ascending"` |
| `descending` | `"sorted descending, activate to sort ascending"` | `"sorted descending"` |
| `sortable` (unsorted) | *(unchanged)* | *(unchanged)* |

Once a user has already sorted a column, they know it's interactive the button role covers that so re-announcing "activate to sort X" every time is repetitive. Just a `default.json` string change, no logic changes needed. Curious what you think.

cc: @Sulaymon333 @jcfilben

Comment (jcfilben) [reply src/js/languages/default.json:80]: I link the idea of shortening the annoucements where possible. I wonder if we run into issues though if the column is pre-sorted via the implementers configuration of the component

Comment (britt6612) [reply src/js/languages/default.json:80]: yes your right I forgot about that case so id be okay then leaving how it is.

**Review by britt6612 — APPROVED** [review]

**Review by jcfilben — COMMENTED** [review]
The 'sortable' message text gets read out both when my focus is on the button and when my focus is on the text itself which is a bit of a weird experience.
Image

**Comment by britt6612** [comment]
> The 'sortable' message text gets read out both when my focus is on the button and when my focus is on the text itself which is a bit of a weird experience.

@jcfilben how did you focus on the text? 
I wonder if the same thing happens with the Cards since were using the similar technique with the `clip: rect(0,0,0,0)` on the `span`

**Comment by jcfilben** [comment]
> @jcfilben how did you focus on the text?
Use ctrl+option+arrow keys or ctrl+option+a

I think the text just needs to be hidden from screenreaders

**Comment by Sulaymon333** [comment]
> The 'sortable' message text gets read out both when my focus is on the button and when my focus is on the text itself which is a bit of a weird experience. Image


@britt6612 I dug into this a bit and we do have similar `HiddenText` implementations across the codebase, but they are wired up differently. `Cards` references its `HiddenText` via `aria-describedby`, and that node sits at the top level, outside any interactive element. Unlike the DataTable Header, where the `HiddenText` is nested inside the `button`.  @jcfilben I tried first using `aria-hidden` on its own actually and it breaks the announcement. So I made the text hidden from screenreaders via `aria-hidden="true"` and ensure announcement is not broken via `aria-labelledby`

Pushed the `aria-hidden` + `aria-labelledby` version via this [commit](https://github.com/grommet/grommet/pull/7988/changes/1d82f1e46406aa7304d905293b6e5337e7cd314a)

**Comment by jcfilben** [comment]
It looks like this change is cause some unexpected chromatic snapshot diffs https://www.chromatic.com/build?appId=5d9774839a6eff00203f5cbf&number=13336

**Comment by Sulaymon333** [comment]
> It looks like this change is cause some unexpected chromatic snapshot diffs https://www.chromatic.com/build?appId=5d9774839a6eff00203f5cbf&number=13336

the snapshot diff is now fixed via this [commit](https://github.com/grommet/grommet/pull/7988/changes/2abdc8044aae39458fff46a9d405976c151c2771)

**Review by jcfilben — APPROVED** [review]
